crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs

```diff
@@ -86,6 +87,59 @@ impl GeneratedAcir {
 }
 
 impl GeneratedAcir {
+    /// Computes lhs mod 2^rhs
+    ///
+    /// `max_bits` is the upper-bound on the bit_size of the object that `lhs` is representing.
+
+    /// An example; max_bits would be 32, if lhs was representing a u32 at a higher level.
+    pub(crate) fn truncate(
+        &mut self,
+        lhs: &Expression,
+        rhs: u32,
+        max_bits: u32,
+    ) -> Result<Expression, AcirGenError> {
+        assert!(max_bits > rhs, "max_bits = {max_bits}, rhs = {rhs}");
+        let exp_big = BigUint::from(2_u32).pow(rhs);
+
+        // 0. Check for constant expression.
+        if let Some(a_c) = lhs.to_const() {
+            let mut a_big = BigUint::from_bytes_be(&a_c.to_be_bytes());
+            a_big %= exp_big;
+            return Ok(Expression::from(FieldElement::from_be_bytes_reduce(&a_big.to_bytes_be())));
+        }
+        let exp = FieldElement::from_be_bytes_reduce(&exp_big.to_bytes_be());
+
+        // 1. Generate witnesses a,b,c
+        let b_witness = self.next_witness_index();
+        let c_witness = self.next_witness_index();
+        self.push_opcode(AcirOpcode::Directive(Directive::Quotient(QuotientDirective {
+            a: lhs.clone(),
+            b: Expression::from_field(exp),
+            q: c_witness,
+            r: b_witness,
+            predicate: None,
+        })));
+
+        self.range_constraint(b_witness, rhs)?;
+        self.range_constraint(c_witness, max_bits - rhs)?;
+
+        // 2. Add the constraint a = b + 2^{rhs} * c
+        //
+        // 2^{rhs}
+        let mut two_pow_rhs_bits = FieldElement::from(2_i128);
+        two_pow_rhs_bits = two_pow_rhs_bits.pow(&FieldElement::from(rhs as i128));
+
+        let b_arith = Expression::from(b_witness);
+        let c_arith = Expression::from(c_witness);
+
+        let res = &b_arith + &(two_pow_rhs_bits * &c_arith);
+        let my_constraint = &res - lhs;
+
+        self.push_opcode(AcirOpcode::Arithmetic(my_constraint));
+
+        Ok(Expression::from(b_witness))
+    }
+
```

Comment:
nit: booleanity cspell (maybe just => boolean)

Reply:
Created #1533 to resolve this